### ISSUE TITLE
Support BART model

### ISSUE THREAD

**Issue by SeolhwaLee** [lxuechen/private-transformers#12]
Hi, 
I'm trying to apply your code in BART model. 
But I got the error like the below:
```
ValueError: Ghost clipping does not support parameter sharing. Parameter sharing may be due to default parameter sharing between lm_head and embedding.Please use a model without parameter sharing for ghost clipping.
```
Does it not support BART model yet??

**Comment by SeolhwaLee** [comment]
Hi, I solved that issue. Thanks! 
But I have another question. I think the main issue for the ghost clipping is duplicate positional embedding.
So I just put the below code in ```modeling_bart.py``` when using ghost clipping. But seems the generation results are super weird. Does it correct way??
```
class BartLearnedPositionalEmbedding(nn.Embedding):
    """
    This module learns positional embeddings up to a fixed maximum size.
    """

    def __init__(self, num_embeddings: int, embedding_dim: int):
        # Bart is set up so that if padding_idx is specified then offset the embedding ids by 2
        # and adjust num_embeddings appropriately. Other models don't have this hack
        self.offset = 2
        super().__init__(num_embeddings + self.offset, embedding_dim)

    def forward(self, input_ids_shape: torch.Size, past_key_values_length: int = 0):
        """`input_ids_shape` is expected to be [bsz x seqlen]."""
        bsz, seq_len = input_ids_shape[:2]
        positions = torch.arange(
            past_key_values_length, past_key_values_length + seq_len, dtype=torch.long, device=self.weight.device
        )
        # --- added: Duplicate to make privacy work! ---
        positions = positions.repeat(bsz, 1)
        # ---
        return super().forward(positions + self.offset)
```

**Comment by lxuechen** [comment]
> I think the main issue for the ghost clipping is duplicate positional embedding. 

That could be one of the issues. Most HF transformers rely on tensor broadcasting for adding positional embedding. This breaks the per-example gradient backward hook. Yeah, the approach you outlined does solve this issue, and it's also what's being done for the other models. 

> But seems the generation results are super weird. Does it correct way?? 

It's very hard to determine what could go wrong w/o further context. If your task is not too complicated, it might be worthwhile to just freeze the positional embedding for the moment and train the other components, e.g.,

```
for param in positional_embedding.parameters():
    param.requires_grad_(False)
```

The privacy engine should be able to ignore parameters that don't require gradients.

**Comment by SeolhwaLee** [comment]
Thank you for your thoughtful advice. I did freeze the positional embeddings as well but seems not to work in my case.
I don't think I need any process for ```generation_utils.py``` for the beam search. But in your code, I saw that you have changed some codebase things on generation utils. Does it relate to my case? P.S. But I have not used your code.

**Comment by lxuechen** [comment]
> But in your code, I saw that you have changed some codebase things on generation utils. 

There shouldn't be major changes there. 

More generally, what are your outputs like? What types of hyperparameters have you used for training? 

If would help me to understand your issue if you could also post the code that was use for training.

**Comment by SeolhwaLee** [comment]
Hi @lxuechen , I made a private repository and invited you. 
So if you take a look at my code then I'm very appreciative :)

**Comment by lxuechen** [comment]
(Partially) addressed in #23.
